Add layout test to ensure the page's scroll position is restored after exiting viewer mode

danae404 opened the pull request:
#### 3288618f90a88dc9a28500d8f62f9eda2c084680
<pre>
Add layout test to ensure the page&apos;s scroll position is restored after exiting viewer mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=295601">https://bugs.webkit.org/show_bug.cgi?id=295601</a>
<a href="https://rdar.apple.com/155372004">rdar://155372004</a>

Reviewed by Jer Noble.

Add a test that checks the scroll position is the same before entering viewer and after exiting viewer.

* LayoutTests/fullscreen/viewer-restore-scroll-position-expected.txt: Added.
* LayoutTests/fullscreen/viewer-restore-scroll-position.html: Added.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::enterViewerMode):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/297210@main">https://commits.webkit.org/297210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0db44f174b72a2e85097d4b4ae7c7b8eca0f23a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110941 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116972 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31284 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39190 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113889 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/24997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/99890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/64796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/24347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/18032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60770 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/18092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37986 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/119773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38362 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/96162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/38186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17893 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37879 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43350 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37542 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40878 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/39247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on LayoutTests/fullscreen/viewer-restore-scroll-position.html:
```diff
@@ -0,0 +1,59 @@
+        waitFor(window, 'load').then(async event => {
+            video = document.getElementsByTagName('video')[0];
+            video.src = findMediaFile('video', '../media/content/test');
+            waitFor(video, 'canplaythrough', true);
```
this should be `await waitFor(...`.

danae404 replied:
Preceding this line with `await` gives me an error: `Unexpected identifier 'waitFor'` 
(adding the other `await`s you suggested do not cause this error)

jernoble replied:
Oh, you can't `waitFor(window, 'load')` because you're not in an `async function` context.

jernoble commented on LayoutTests/fullscreen/viewer-restore-scroll-position.html:
```diff
@@ -0,0 +1,59 @@
+            video.src = findMediaFile('video', '../media/content/test');
+            waitFor(video, 'canplaythrough', true);
+            
+            runWithKeyDown(() => video.play());
```
`await runWithKeyDown(`.

jernoble commented on LayoutTests/fullscreen/viewer-restore-scroll-position.html:
```diff
@@ -0,0 +1,59 @@
+            waitFor(video, 'canplaythrough', true);
+            
+            runWithKeyDown(() => video.play());
+            waitFor(video, 'play', true);
```
`await waitFor(...`.

jernoble commented on LayoutTests/fullscreen/viewer-restore-scroll-position.html:
```diff
@@ -0,0 +1,59 @@
+                    testRunner.notifyDone();
+            };
+
+            runWithKeyDown(() => window.internals.enterViewerMode(video));
```
You could do something like this:

```suggestion
            runWithKeyDown(() => window.internals.enterViewerMode(video));
            await  waitFor(document, 'fullscreenchange');

            await sleepFor(100);
            document.webkitCancelFullScreen();
            
            await waitFor(document, 'fullscreenchange')
            await testExpectedEventually("(document.scrollingElement.scrollTop === originalScroll)", true);
```

jernoble replied:
I think this will significantly change the -expected.txt file though, because you're not logging most of the events you `waitFor` here.

danae404 replied:
I posted a new patch with your suggestions, except in order to get the test to run without timing out I did not add `await` before `waitFor(window, 'load').then(async event => {` and I also kept                 
```
if (window.testRunner)
    testRunner.notifyDone(); 
```

danae404 committed:
Add layout test to ensure the page's scroll position is restored after exiting viewer mode
https://bugs.webkit.org/show_bug.cgi?id=295601
rdar://155372004

Reviewed by Jer Noble.

Add a test that checks the scroll position is the same before entering viewer and after exiting viewer.

* LayoutTests/fullscreen/viewer-restore-scroll-position-expected.txt: Added.
* LayoutTests/fullscreen/viewer-restore-scroll-position.html: Added.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::enterViewerMode):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/297210@main
